wkcuber/api/Layer.py

```diff
@@ -99,20 +118,157 @@ def set_bounding_box_size(self, size: Tuple[int, int, int]) -> None:
         for _, mag in self.mags.items():
             mag.view.size = size
 
-    def setup_mag(self, mag: str) -> None:
+    def _initialize_mag_from_other_mag(
+        self, new_mag_name: Union[str, Mag], other_mag: MagDataset, compress: bool
+    ) -> MagDataset:
+        raise NotImplemented
+
+    def downsample(
+        self,
+        from_mag: Mag,
+        max_mag: Mag,
+        interpolation_mode: str,
+        compress: bool,
+        scale: Tuple[float, float, float] = None,
+        buffer_edge_len: int = None,
+        args: Namespace = None,
+    ) -> None:
+        # if 'scale' is set, the data gets downsampled anisotropic
+
+        # pad all existing mags if necessary
+        # during each downsampling step, the data shape or offset of the new mag should never need to be rounded
+        existing_mags = sorted([Mag(mag) for mag in self.mags.keys()])
+        all_mags_after_downsampling = existing_mags.copy()
+        cur_mag = get_next_mag(from_mag, scale)
+        while cur_mag <= max_mag:
+            all_mags_after_downsampling += [cur_mag]
+            cur_mag = get_next_mag(cur_mag, scale)
+
+        lowest_mag = all_mags_after_downsampling[-1].as_np()
+
+        # calculate aligned offset and size for lowest mag
+        offset_in_mag1 = self.dataset.properties.data_layers[
+            self.name
+        ].get_bounding_box_offset()
+        size_in_mag1 = self.dataset.properties.data_layers[
+            self.name
+        ].get_bounding_box_size()
+        offset_in_lowest_mag = offset_in_mag1 // all_mags_after_downsampling[-1].as_np()
+        end_offset_in_lowest_mag = -(
+            (np.array(offset_in_mag1) + size_in_mag1)
+            // -all_mags_after_downsampling[-1].as_np()
+        )  # ceil div
+        # translate alignment into mag 1
+        aligned_offset_in_mag1 = lowest_mag * offset_in_lowest_mag
+        aligned_size_in_mag1 = lowest_mag * (
+            end_offset_in_lowest_mag - offset_in_lowest_mag
+        )
+
+        # pad the existing mags
+        for mag_name in existing_mags:
+            mag = self.mags[mag_name.to_layer_name()]
+            aligned_offset = aligned_offset_in_mag1 // mag_name.as_np()
+            aligned_size = aligned_size_in_mag1 // mag_name.as_np()
+            current_offset = mag.get_view().global_offset
+            current_size = mag.get_view().size
+
+            shape = list((self.num_channels,) + tuple(aligned_size))
+            # pad (left + right) and (top + bottom) and (front + back)
+            for i in range(0, 3):
+                # pad left / top / front
+                buffer_width = (np.array(current_offset) - aligned_offset)[i]
+                if buffer_width > 0:
+                    padding_shape = shape
+                    padding_shape[i + 1] = buffer_width
+                    mag.write(
+                        data=np.zeros(padding_shape, dtype=mag.get_dtype()),
+                        offset=aligned_offset,
+                    )
```

Comment:
I assume this only writes zeros, so that the bounding box is enlarged as a side-effect? Two questions:

- can we avoid writing zeros? (a wkw should return zeros anyway if it the target area is empty)
- should this be undone after the downsampling is finished so that the bounding box stays tight?

Reply:
1. I haven't tested this, but I assume that it would be sufficient for `WKDataset` to update the boundingbox (without padding the data) (because, as you pointed out, wkw handles this). ~~ I think this would not work for `TiffDataset`.~~ The `TiffDataset` does not add padding during reading, but during writing.
2. This cannot be undone. The amount of padding is dependent on the lowest magnification (so that for every two adjacent magnifications the ratio of their offsets and sizes is equal to `scale`). Resetting the boundingbox would result in rounding problems in lower mags

Reply:
Only updating the properties does work. I have tested it locally, but before merging we should test everything on a large dataset and look at the results in webknossos. The large dataset should have an offset, so that as many edge cases as possible are covered (like this 'padding').